### ISSUE TITLE
np.asarray returns non-writable array

### ISSUE THREAD

**Issue by ericpre** [dask/dask#6520]
**What happened**:
Behaviour change in the last release (2.23.0): converting a dask array to a numpy array results in a non-writable array when it has been created with `np.zeros_like` of a dask array.

**What you expected to happen**:
The returned numpy array should be consistently writable (or not?), regardless if the dask array have been created with `da.arange` or `da.zeros`, for example.

**Minimal Complete Verifiable Example**:

```python
import numpy as np
import dask.array as da

dask_array = np.zeros_like(da.arange(20))
numpy_array = np.asarray(dask_array)
numpy_array[0] = 1
```
Error:
```python
Traceback (most recent call last):

  File "<ipython-input-19-6eef1a4f1c7b>", line 5, in <module>
    numpy_array[0] = 1

ValueError: assignment destination is read-only
```

The following example works fine:
```python
dask_array = da.arange(20)
numpy_array = np.asarray(dask_array)
numpy_array[0] = 1
```

Another example, which is not working with dask 0.23.0:
```python
dask_array = da.zeros((20))
numpy_array = np.asarray(dask_array)
numpy_array[0] = 1
```

**Anything else we need to know?**:

Checking the numpy flags, it seems that the numpy array is not writable anymore.

**Environment**:

- Dask version: 2.23.0
- Python version: 3.7
- Operating System: Fedora
- Install method (conda, pip, source): conda

**Comment by mrocklin** [comment]
cc @jakirkham

**Comment by jsignell** [comment]
I did a little detective work and am fairly confident that this is related to #6491. If this is a supported behavior, then we should add some tests and perhaps revert that PR. @Carreau in case you have other ideas on how to proceed.

**Comment by Carreau** [comment]
Ah, yes this is likely due to that PR, I though they might be cases where this would trigger, but couldn't find one. It's likely safest to revert for now to not break user code and think as to wether this is supported or how to work around it.

**Comment by jsignell** [comment]
Can someone on @dask/maintenance confirm that this is a behavior that we intend to support?

**Comment by jsignell** [comment]
Agreed. Thanks for providing that diff.

**Comment by jakirkham** [comment]
Agree with the consensus this is related to PR ( https://github.com/dask/dask/pull/6491 ). Certainly one option (if agreeable) is to revert. Another option may be to ensure results from `.compute()` are writeable, which would fix this issue.

**Comment by Carreau** [comment]
#6525 as a suggestion; making result of `.compute()` writable would mean the following would not be true:

```
import dask.array as da
import numpy as np
dask_array = np.zeros_like(da.arange(20000))
d = dask_array.compute()
d.base.nbytes == 8
```

It's up to you.

**Comment by jakirkham** [comment]
Sorry I'm not following. What are we concerned about here?

**Comment by Carreau** [comment]
Number of bytes of the return value; the goal of #6491 was to keep arrays views to the same value small when possible; if we copy in `finalize()`, then we always end up back with a full numpy array that have many copies of the value.

**Comment by jakirkham** [comment]
I thought the goal of that PR was to cutdown on communication time.

Anyways `finalize` performs concatenation of potentially many smaller arrays from chunks. IOW it typically copies the data. So would expect the resulting array to take up as much space as one would naively expected (sans tricks).

**Comment by Carreau** [comment]
> I thought the goal of that PR was to cutdown on communication time.

> Anyways finalize performs concatenation of potentially many smaller arrays from chunks. IOW it typically copies the data. So would expect the resulting array to take up as much space as one would naively expected (sans tricks).

Yeah, I would have hoped to potentially keep the `.compute()` small if possible.

Do you think a copy on write patch would also be feasible ? like only on `Arrays.__setitem__`, and `__array__` ?

**Comment by Carreau** [comment]
> Do you think a copy on write patch would also be feasible ? like only on `Arrays.__setitem__`, and `__array__` ?

No it would not, unless `__array__` returns a Non-numpy array with is copy on write.

**Comment by shoyer** [comment]
I don't think it's a good idea to force computing a dask array to return a writeable array from `np.asarray()` by copying. That imposes a performance burden for all users based on a rather niche use-case.

If you want to ensure a writeable array, you can use `np.array()` (which always makes a copy) instead of `np.asarray()`

**Comment by jakirkham** [comment]
So `np.asarray` already has the behavior of [computing a Dask Array]( https://github.com/dask/dask/blob/00843e3f810b0b76a8e71f97f4825ab616d9296b/dask/array/core.py#L1364 ) and is used in the wild for that purpose. I don't think that is what is being discussed here.

I think the only question here is, when a Dask Array is computed, should the resulting array be mutable? Other than the fact that a few cases (like where a single chunk constitutes the full Dask Array) will require a copy (the bulk of cases won't), I'm not sure I see a reason we can't guarantee this for the end user (especially as that was true in recent history).

That said, if that's not agreeable, the user can always do `np.require(a, requirements="W")`.

**Comment by jakirkham** [comment]
> I think the only question here is, when a Dask Array is computed, should the resulting array be mutable? Other than the fact that a few cases (like where a single chunk constitutes the full Dask Array) will require a copy (the bulk of cases won't), I'm not sure I see a reason we can't guarantee this for the end user (especially as that was true in recent history).

Hmm...missed Stephan's comment ( https://github.com/dask/dask/pull/6525#issuecomment-677084228 ). Yeah I can see how this may be tricky to solve in a way that doesn't add more footguns.

**Comment by Carreau** [comment]
I think there is also two considerations here:
  - What is the behavior we want in the long term
  - Are we breaking user code _now_.

I agree with Stephan's comment and I am in the mind that dask arrays should not be writable unless explicitely asked to for some for the reason he underlines.

Though I think any changes in this behavior should be properly decided discuss and go through a deprecation warning process to give a chance for user to update their code. Otherwise they will just ping dask, and never upgrade as they see this change as a bug.

**Comment by ericpre** [comment]
> I don't think it's a good idea to force computing a dask array to return a writeable array from `np.asarray()` by copying. That imposes a performance burden for all users based on a rather niche use-case.
> 
> If you want to ensure a writeable array, you can use `np.array()` (which always makes a copy) instead of `np.asarray()`

This sounds sensible as long term expected behaviour.

> I think there is also two considerations here:
> 
> * What is the behavior we want in the long term
> * Are we breaking user code _now_.

Difficult to tell, but if this was breaking the code of many people, we would expect this issue to be reference to other issues?! At the moment, I am the only one to have mentioned it!

Maybe it simply need to be documented?

**Comment by Carreau** [comment]
> Difficult to tell, but if this was breaking the code of many people, we would expect this issue to be reference to other issues?! At the moment, I am the only one to have mentioned it!

Not necessarily, because of 1) delay of adoption, and 2) corporate might not be filling public bugs.

I think the time-from-release-to-bug-report is an indicator of how broadly it could be relied upon.
